mrp_bom_current_stock/wizard/bom_route_current_stock.py

```diff
@@ -0,0 +1,155 @@
+            'bom_line': bom_line.id,
+            'bom_level': level,
+            'product_qty': bom_line.product_qty * factor,
+            'product_uom_id': bom_line.product_uom.id,
+            'location_id': (
+                bom_line.location_id.id if bom_line.location_id else False,
+            ),
+            'explosion_id': self.id,
+        }
+
+    @api.multi
+    def do_explode(self):
+        self.ensure_one()
+        line_obj = self.env['mrp.bom.current.stock.line']
+
+        def _create_lines(bom, level=0, factor=1):
+            level += 1
+            for line in bom.bom_line_ids:
+                vals = self._prepare_line(line, level, factor)
+                line_obj.create(vals)
+                location = line.location_id
+                line_boms = line.product_id.bom_ids
+                bom = line_boms.filtered(
+                    lambda bom: bom.location_id == location
+                ) or line_boms.filtered(lambda b: not b.location_id)
+                if bom:
+                    line_qty = self.env['product.uom']._compute_qty_obj(
+                        line.product_uom, line.product_qty, bom.product_uom
+                    )
+                    new_factor = factor * line_qty / bom.product_qty
+                    _create_lines(bom[0], level, new_factor)
+
+        _create_lines(self.bom_id)
+        return {
+            'type': 'ir.actions.act_window',
+            'name': 'Open lines',
+            'view_mode': 'form',
+            'view_type': 'form',
+            'res_model': 'mrp.bom.current.stock',
+            'view_id': self.env.ref(
+                'mrp_bom_current_stock.mrp_bom_current_stock_view_form2'
+            ).id,
+            'target': 'new',
+            'res_id': self.id,
+        }
+
+
+class BomRouteCurrentStockLine(models.TransientModel):
+    _name = "mrp.bom.current.stock.line"
+
+    explosion_id = fields.Many2one(
+        comodel_name='mrp.bom.current.stock',
+    )
+    product_id = fields.Many2one(
+        comodel_name='product.product',
+        string='Product Variant',
+        readonly=True,
+    )
+    bom_level = fields.Integer(
+        string='BoM Level',
+        readonly=True
+    )
+    product_qty = fields.Float(
+        string='Product Quantity',
+        readonly=True,
+        digits=dp.get_precision('Product Unit of Measure'),
+    )
+    product_uom_id = fields.Many2one(
+        comodel_name='product.uom',
+        string='Product Unit of Measure',
+        readonly=True,
+    )
+    location_id = fields.Many2one(
+        comodel_name="stock.location",
+        string="Source location",
+    )
+    bom_line = fields.Many2one(
+        comodel_name="mrp.bom.line",
+        string="BoM line",
+    )
+    qty_available_in_source_loc = fields.Float(
+        string="Qty Available in Source",
+        compute="_compute_qty_available_in_source_loc",
+    )
+
+    def _compute_qty_available_in_source_loc(self):
+        self.ensure_one()
```

Comment:
This should be api.multi with a for loop.
Otherwise, we get an error printing the report with multiple bom.lines :
```
  File "/opt/odoo/src/odoo/fields.py", line 998, in _compute_value
    getattr(records, self.compute)()
  File "/opt/odoo/external-src/manufacture-reporting/mrp_bom_current_stock/wizard/bom_route_current_stock.py", line 148, in _compute_qty_available_in_source_loc
    self.ensure_one()
  File "/opt/odoo/src/odoo/models.py", line 4283, in ensure_one
    raise ValueError("Expected singleton: %s" % self)
ValueError: Expected singleton: mrp.bom.current.stock.line(2, 3, 4)

Error to render compiling AST
ValueError: Expected singleton: mrp.bom.current.stock.line(2, 3, 4)
Template: 1259
Path: /templates/t/t/t/div/t/table/tbody/tr[2]/td[5]/span[1]
Node: <span t-field="l.qty_available_in_source_loc"/>
```